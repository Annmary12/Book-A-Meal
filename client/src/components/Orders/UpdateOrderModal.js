import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactModal from 'react-modal';
import OrderInfoForm from './OrderInfoForm.js';

/**
 * @class UpdateOrderModal
 */
class UpdateOrderModal extends Component {
  static propTypes = {
    isOpen: PropTypes.bool.isRequired,
    handleClose: PropTypes.func.isRequired,
    handleUpdateOrder: PropTypes.func.isRequired,
    order: PropTypes.shape({
      id: PropTypes.string.isRequired,
      meal: PropTypes.shape({
        id: PropTypes.string.isRequired,
        img: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired,
        price: PropTypes.number.isRequired
      }).isRequired,
      deliveryAddress: PropTypes.string.isRequired,
      phoneNumber: PropTypes.string.isRequired,
      quantity: PropTypes.number.isRequired
    }).isRequired
  }

  state = {
    updateOrderRequestIsProcessing: false
  };

  handleClose = () => {
    this.setState({
      updateOrderRequestIsProcessing: false
    });
    this.props.handleClose();
  }


  updateOrder = (orderDetails) => {
    const { order, handleUpdateOrder } = this.props;

    this.setState({ updateOrderRequestIsProcessing: true });

    return handleUpdateOrder(order.id, orderDetails)
      .then(this.handleClose);
  }

  /**
   * @return {JSX} React jsx
   */
  render() {
    const {
      isOpen, handleClose, order
    } = this.props;

    return (
      <ReactModal
        isOpen={isOpen}
        onRequestClose={handleClose}
        overlayClassName="react-modal-overlay"
        className="react-modal-body"
        closeTimeoutMS={150}
      >
        <OrderInfoForm
          meal={order.meal}
          quantity={order.quantity}
          deliveryAddress={order.deliveryAddress}
          phoneNumber={order.phoneNumber}
          handleSubmit={this.updateOrder}
          proceedText="Confirm update"
          headerText="Confirm order information"
          isProcessing={this.state.updateOrderRequestIsProcessing}
          handleClose={handleClose}
        />
      </ReactModal>
    );
  }
}

export default UpdateOrderModal;
