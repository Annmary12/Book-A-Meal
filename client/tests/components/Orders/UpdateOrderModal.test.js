import React from 'react';
import UpdateOrderModal from '../../../src/components/Orders/UpdateOrderModal';

const props = {
  isOpen: false,
  handleClose: jest.fn(),
  handleUpdateOrder: (orderId, orderDetails) => Promise.resolve([orderId, orderDetails]),
  order: {
    id: 'asdf23e',
    meal: {
      id: 'af328d',
      img: 'picture.jpg',
      title: 'garri and sugar',
      description: 'Garri and sugar with enough water',
      price: 230
    },
    deliveryAddress: 'adf23 sadf829e3r',
    phoneNumber: '08123324234',
    quantity: 2
  }
}
describe('UpdateOrderModal component', () => {
  const wrapper = shallow(
    <UpdateOrderModal {...props} />
  );

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('handleClose method', () => {
    wrapper.instance().handleClose();

    it('should set updateOrderRequestIsProcessing is false', () => {
      const { updateOrderRequestIsProcessing } = wrapper.instance().state;

      expect(updateOrderRequestIsProcessing).toBeFalsy();
    });

    it('should call handleClose function from prop', () => {
      expect(props.handleClose).toBeCalled();
    });
  });

  describe('updateOrder method', () => {
    it('should set updateOrderRequestIsProcessing to true', async () => {
      const orderDetails = { id: 'tdf82fsadf'};
      const action = wrapper.instance();

      action.updateOrder(orderDetails);
      const { updateOrderRequestIsProcessing } = wrapper.state();

      expect(updateOrderRequestIsProcessing).toBeTruthy();
    });
  });
});