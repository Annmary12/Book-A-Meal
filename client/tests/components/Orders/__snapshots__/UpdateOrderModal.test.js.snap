// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateOrderModal component should match snapshot 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="react-modal-body"
  closeTimeoutMS={150}
  isOpen={false}
  onRequestClose={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  overlayClassName="react-modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <OrderInfoForm
    deliveryAddress="adf23 sadf829e3r"
    handleClose={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    handleSubmit={[Function]}
    headerText="Confirm order information"
    isProcessing={false}
    meal={
      Object {
        "description": "Garri and sugar with enough water",
        "id": "af328d",
        "img": "picture.jpg",
        "price": 230,
        "title": "garri and sugar",
      }
    }
    phoneNumber="08123324234"
    proceedText="Confirm update"
    quantity={2}
  />
</Modal>
`;
